challenge/files/12_context.toml

```diff
@@ -0,0 +1,103 @@
+name = "Context"
+category = "easy"
+default_code_snippet = """
+var (
+	errTimeout error = errors.New("Timeout")
+)
+
+func main() {
+	// ===============================
+	// === Code for questions 1, 2 ===
+	// ===============================
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	timeoutCtx, timeoutCancel := context.WithTimeoutCause(context.Background(), 2*time.Second, errTimeout)
+	deadlineCtx, deadlineCancel := context.WithDeadline(context.Background(), time.Now().Add(2*time.Second))
+	valueCtx := context.WithValue(cancelCtx, "Key", "Value")
+	time.AfterFunc(2*time.Second, func() {
+		cancel()
+	})
+
+	defer func() {
+		cancel()
+		timeoutCancel()
+		deadlineCancel()
+	}()
+
+	go func() {
+		<-timeoutCtx.Done()
+		fmt.Println("Timeout context")
+		fmt.Printf("%v\\n", context.Cause(timeoutCtx))
+		fmt.Printf("%v\\n", timeoutCtx.Err())
+	}()
+
+	go func() {
+		<-deadlineCtx.Done()
+		fmt.Println("Deadline context")
+	}()
+
+	go func() {
+		<-valueCtx.Done()
+		fmt.Println("Child context")
+		fmt.Println(valueCtx.Value("Key"))
+	}()
+
+	time.Sleep(3 * time.Second)
+
+	// ===============================
+	// === Code for questions 3 ===
+	// ===============================
+
+	ctx1, _ := context.WithTimeout(context.Background(), 1*time.Second)
+	ctx2, _ := context.WithTimeout(ctx1, 5*time.Second)
+
+	ctx4, _ := context.WithTimeout(context.Background(), 5*time.Second)
+	ctx3, _ := context.WithTimeout(ctx4, 1*time.Second)
+
+	_, _ = ctx2, ctx4
+}
+"""
+learning_advise = """
+“You must be the change you wish to see in the world.” – Mahatma Gandhi.
+"""
+
+[[learning_links]]
+title = "Go package 'context'"
+url = "https://pkg.go.dev/context"
+
+[[learning_links]]
+title = "Go by Example: Context"
+url = "https://gobyexample.com/context"
+
+[[learning_links]]
+title = "How To Use Contexts in Go"
+url = "https://www.digitalocean.com/community/tutorials/how-to-use-contexts-in-go"
+
+[[questions]]
+text = """Which context will be cancelled first?"""
+type = "select_answers"
+answers = [
+    { text = "Timeout"},
+    { text = "Deadline"},
+    { text = "Value"},
+    { text = "They will be cancelled at the same time", code_line_ranges = []  },
+]
+
+[[questions]]
+text = """What will be the output of the code on lines 25-26?"""
+type = "select_answers"
+answers = [
+    { text = """Timeout\nTimeout"""},
+    { text = """context deadline exceeded\ncontext deadline exceeded"""},
+    { text = """Timeout\ncontext deadline exceeded""", code_line_ranges = []  },
+    { text = """context deadline exceeded\nTimeout"""},
+]
+
+[[questions]]
+text = """Choose the right match between context and cancellation timeout"""
```

Comment:
Let's ask about this code. What program prints: 1,2,3,5

```go
	start := time.Now()

	ctx1, _ := context.WithTimeout(context.Background(), 3*time.Second)
	ctx2, _ := context.WithTimeout(ctx1, 2*time.Second)
	ctx3, _ := context.WithTimeout(ctx2, 5*time.Second)

	<-ctx3.Done()

	fmt.Println(time.Since(start).Round(time.Second).Seconds())
```